lib/types.nix

```diff
@@ -109,24 +110,30 @@ rec {
 
     listOf = elemType: mkOptionType {
       name = "list of ${elemType.name}s";
-      check = value: isList value && all elemType.check value;
+      check = isList;
```

Comment:
I'm sorry, but why the hell was this done? Please explain otherwise I am going to revert this. (Same for all other changed `check` functions)

Reply:
I think one reason might be that, with this change, we get more localised errors, for example
```
A definition for option `foo.[definition 1-entry 2]' is not of type `bar'.
```
instead of
```
A definition for option `foo' is not of type `list of bar'.
```
and similarly for `attrsOf`. I think we want to keep this behaviour.

We should also be careful not to duplicate work, as type checking is currently done by `mergeDefinitions`.

Reply:
(And whatever we do, this *really* needs more documentation, for example the manual doesn't mention that `type.check` only checks one "layer" of types.)

Reply:
Thanks for the input. How about type checking the `meta` attribute as an entire module system like in NixOS, instead of doing things manually?

```nix
      attrs = (lib.evalModules {
        modules = [ {
          options = metaTypes;
          config = meta;
        } ];
      }).config;
```

This allows us to leverage all of the existing infrastructure, and the only thing that needs to change for that is to modify `metaTypes` to become a "module" instead of just a type declaration.

(Alternatively, how would `checkType` look like? How do I even use `mergeDefinitions`?)

Reply:
> I'm sorry, but why the hell was this done? Please explain otherwise I am going to revert this. (Same for all other changed `check` functions)

Previous attempts at fixing this:

- https://github.com/NixOS/nixpkgs/pull/173568
- https://github.com/NixOS/nixpkgs/pull/168295

I think there are several others.